src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1215,59 +1227,80 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
         {
             Debug.Assert(!IsConditionalState);
 
-            VisitRvalue(node.LeftOperand);
-            if (IsConstantNull(node.LeftOperand))
+            BoundExpression leftOperand = node.LeftOperand;
+            BoundExpression rightOperand = node.RightOperand;
+
+            Result leftResult = VisitRvalueWithResult(leftOperand);
+            if (IsConstantNull(leftOperand))
             {
-                VisitRvalue(node.RightOperand);
+                VisitRvalue(rightOperand);
                 return null;
             }
 
             var leftState = this.State.Clone();
-            var leftType = _result.Type;
-            if (leftType?.IsNullable == false)
+            if (leftResult.Type?.IsNullable == false)
             {
-                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, node.LeftOperand.Syntax);
+                ReportStaticNullCheckingDiagnostics(ErrorCode.HDN_ExpressionIsProbablyNeverNull, leftOperand.Syntax);
             }
 
-            bool leftIsConstant = node.LeftOperand.ConstantValue != null;
+            bool leftIsConstant = leftOperand.ConstantValue != null;
             if (leftIsConstant)
             {
                 SetUnreachable();
             }
 
-            leftType = InferResultNullability(node.LeftOperand, node.LeftConversion, node.Type, leftType);
-
-            VisitRvalue(node.RightOperand);
-            var rightType = _result.Type;
-
+            var rightResult = VisitRvalueWithResult(rightOperand);
```

Comment:
>var rightResult = VisitRvalueWithResult(rightOperand); [](start = 12, length = 54)

It feels that for the case when result type is driven by the left side we should be unwrapping implicit conversions on top of the right operand and then reapplying them back.

Reply:
Added `// PROTOTYPE` comment.